src/main/java/org/opensearch/security/http/OnBehalfOfAuthenticator.java

```diff
@@ -0,0 +1,253 @@
+        final SecurityManager sm = System.getSecurityManager();
+
+        if (sm != null) {
+            sm.checkPermission(new SpecialPermission());
+        }
+
+        AuthCredentials creds = AccessController.doPrivileged(new PrivilegedAction<AuthCredentials>() {
+            @Override
+            public AuthCredentials run() {
+                return extractCredentials0(request);
+            }
+        });
+
+        return creds;
+    }
+
+    private AuthCredentials extractCredentials0(final RestRequest request) {
+        if (!oboEnabled) {
+            log.error("On-behalf-of authentication has been disabled");
+            return null;
+        }
+
+        if (jwtParser == null) {
+            log.error("Missing Signing Key. JWT authentication will not work");
+            return null;
+        }
+
+        String jwtToken = request.header(HttpHeaders.AUTHORIZATION);
+
+        if (jwtToken == null || jwtToken.length() == 0) {
+            if (log.isDebugEnabled()) {
+                log.debug("No JWT token found in '{}' header", HttpHeaders.AUTHORIZATION);
+            }
+            return null;
+        }
+
+        if (!BEARER.matcher(jwtToken).matches()) {
+            jwtToken = null;
+        }
+
+        if (jwtToken != null && Pattern.compile(BEARER_PREFIX).matcher(jwtToken.toLowerCase()).find()) {
+            jwtToken = jwtToken.substring(jwtToken.toLowerCase().indexOf(BEARER_PREFIX) + BEARER_PREFIX.length());
+        } else {
+            if (log.isDebugEnabled()) {
+                log.debug("No Bearer scheme found in header");
+            }
+        }
+
+        if (jwtToken == null) {
+            return null;
+        }
+
+        try {
+            Matcher matcher = PATTERN_PATH_PREFIX.matcher(request.path());
+            final String suffix = matcher.matches() ? matcher.group(2) : null;
+            if (request.method() == RestRequest.Method.POST && ON_BEHALF_OF_SUFFIX.equals(suffix)
+                || request.method() == RestRequest.Method.PUT && ACCOUNT_SUFFIX.equals(suffix)) {
+                final OpenSearchException exception = ExceptionUtils.invalidUsageOfOBOTokenException();
+                log.error(exception.toString());
+                return null;
+            }
+
+            final Claims claims = jwtParser.parseClaimsJws(jwtToken).getBody();
+
+            final String subject = claims.getSubject();
+            if (Objects.isNull(subject)) {
+                log.error("Valid jwt on behalf of token with no subject");
+                return null;
+            }
+
+            final String audience = claims.getAudience();
+            if (Objects.isNull(audience)) {
+                log.error("Valid jwt on behalf of token with no audience");
+                return null;
+            }
+
+            final String tokenType = claims.get(TOKEN_TYPE_CLAIM).toString();
+            if (!tokenType.equals(TOKEN_TYPE)) {
```

Comment:
nit: token

Does this mean that an OBO token must have a claim like `type: obo` to be an on-behalf-of token even if its signed with the right key?

IMO It makes sense to include `type` in the token if it triggers logic specific to the type, but otherwise I'm not sure its needed and takes up space in the token.